File: challenges/url-shortener/python/solution/exercise.py
"""Reference: reversible allocation through two maps."""


class URLShortener:
    def __init__(self):
        self._urls: dict[str, str] = {}
        self._codes: dict[str, str] = {}
        self._next_id = 1

    def shorten(self, url: str) -> str:
        if url in self._urls:
            return self._urls[url]
        alphabet = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
        number = self._next_id
        digits = []
        while number:
            number, remainder = divmod(number, 62)
            digits.append(alphabet[remainder])
        code = "".join(reversed(digits))
        self._next_id += 1
        self._urls[url] = code
        self._codes[code] = url
        return code

    def resolve(self, code: str) -> str | None:
        return self._codes.get(code)

File: challenges/url-shortener/typescript/solution/exercise.ts
export class URLShortener {
  private urls = new Map<string, string>();
  private codes = new Map<string, string>();
  private nextId = 1;

  shorten(url: string): string {
    const existing = this.urls.get(url);
    if (existing !== undefined) return existing;
    const alphabet = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    let number = this.nextId++;
    const digits: string[] = [];
    while (number > 0) {
      digits.push(alphabet[number % 62]);
      number = Math.floor(number / 62);
    }
    const code = digits.reverse().join("");
    this.urls.set(url, code);
    this.codes.set(code, url);
    return code;
  }

  resolve(code: string): string | null {
    return this.codes.get(code) ?? null;
  }
}
